pkg/qrcodegenerator/grpc_gateways_test.go

```diff
@@ -0,0 +1,130 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package qrcodegenerator_test
+
+import (
+	"testing"
+
+	"github.com/smarty/assertions"
+	"go.thethings.network/lorawan-stack/v3/pkg/component"
+	componenttest "go.thethings.network/lorawan-stack/v3/pkg/component/test"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	. "go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator"
+	"go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator/qrcode/gateways"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/util/test"
+	"go.thethings.network/lorawan-stack/v3/pkg/util/test/assertions/should"
+)
+
+func TestGatewayQRCodeParsing(t *testing.T) {
+	a := assertions.New(t)
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+
+	c := componenttest.NewComponent(t, &component.Config{})
+	ttigpro1 := new(gateways.TTIgpro1Format)
+	qrg, err := New(c, &Config{}, WithGatewayFormat(ttigpro1.ID(), ttigpro1))
+	test.Must(qrg, err)
+
+	componenttest.StartComponent(t, c)
+	defer c.Close()
+
+	mustHavePeer(ctx, c, ttnpb.ClusterRole_QR_CODE_GENERATOR)
+
+	client := ttnpb.NewGatewayQRCodeGeneratorClient(c.LoopbackConn())
+
+	for _, tc := range []struct {
+		Name      string
+		FormatID  string
+		GetQRData func() []byte
+		Assertion func(*assertions.Assertion, *ttnpb.ParseGatewayQRCodeResponse, error) bool
+	}{
+		{
+			Name: "EmptyData",
+			GetQRData: func() []byte {
+				return []byte{}
+			},
+			Assertion: func(a *assertions.Assertion, resp *ttnpb.ParseGatewayQRCodeResponse, err error) bool {
+				if !a.So(resp, should.BeNil) {
+					return false
+				}
+				if !a.So(errors.IsInvalidArgument(err), should.BeTrue) {
+					return false
+				}
+				return true
+			},
+		},
+		{
+			Name:     "UnknownFormat",
+			FormatID: "unknown",
+			GetQRData: func() []byte {
+				return []byte(`https://ttig.pro/c/ec656efffe000128/abcdef123456`)
+			},
+			Assertion: func(a *assertions.Assertion, resp *ttnpb.ParseGatewayQRCodeResponse, err error) bool {
+				if !a.So(resp, should.BeNil) {
+					return false
+				}
+				if !a.So(errors.IsInvalidArgument(err), should.BeTrue) {
+					return false
+				}
+				return true
+			},
+		},
+		{
+			Name:     "InvalidFormat",
+			FormatID: "tr005",
+			GetQRData: func() []byte {
+				return []byte(`https://ttig.pro/c/ec656efffe000128/abcdef123456`)
+			},
+			Assertion: func(a *assertions.Assertion, resp *ttnpb.ParseGatewayQRCodeResponse, err error) bool {
+				if !a.So(resp, should.BeNil) {
+					return false
+				}
+				if !a.So(errors.IsInvalidArgument(err), should.BeTrue) {
+					return false
+				}
+				return true
+			},
+		},
+		{
+			Name:     "ValidTTIgpro1",
```

Comment:
Casing is `TTIGPRO1` written, ID is `ttigpro1`

Reply:
fixed.